File: packages/sdk/src/resources/sessions.ts
import type { DocupleteClient } from "../client.js";
import type {
  Session,
  SessionListItem,
  CreateSessionParams,
  CreateSessionResult,
  ListSessionsParams,
  ListSessionsResult,
  GenerateSessionResult,
  GenerateStatusResult,
  BulkCreateSessionItem,
  BulkCreateSessionResult,
  AuditLogResult,
  SessionSignersResult,
} from "../types.js";

interface GetSessionResponse {
  session: Session;
}

export interface SendLinkParams {
  /** Required. The recipient's email address. */
  recipientEmail: string;
  /** Optional. The recipient's display name used in the email greeting. */
  recipientName?: string;
  /** Optional. A custom message appended to the standard email body. */
  customMessage?: string;
}

export interface VoidSessionParams {
  /** Optional reason for voiding, stored in the audit log. */
  reason?: string;
  /** When `true`, notifies the signer by email that the session was voided. Default: false. */
  notifySigner?: boolean;
}

export class SessionsResource {
  constructor(private readonly client: DocupleteClient) {}

  /**
   * List sessions for your account with optional filters.
   * Returns sessions in descending `created_at` order.
   *
   * @example
   * ```ts
   * // List all submitted sessions for a specific package
   * const { sessions, total } = await client.sessions.list({
   *   packageId: 42,
   *   status: "generated",
   *   limit: 50,
   * });
   *
   * // Incremental sync — only sessions updated since last poll
   * const { sessions } = await client.sessions.list({
   *   updatedAfter: "2024-01-15T00:00:00Z",
   * });
   * ```
   */
  async list(params: ListSessionsParams = {}): Promise<ListSessionsResult> {
    const query: Record<string, string | number | undefined> = {
      packageId:    params.packageId,
      status:       params.status,
      limit:        params.limit,
      offset:       params.offset,
      updatedAfter: params.updatedAfter,
      search:       params.search,
    };
    return this.client.get<ListSessionsResult>("/sessions", query);
  }

  /**
   * Create a new interview session via the headless API.
   *
   * Returns a `sessionToken` and a ready-to-use `interviewUrl` your client
   * can be sent to or redirected at.
   *
   * Prefill values should use field **source keys** as keys
   * (e.g. `{ firstName: "Jane", email: "jane@example.com" }`).
   *
   * Requires a `dp_live_…` API key.
   *
   * @example
   * ```ts
   * const { sessionToken, interviewUrl } = await client.sessions.create({
   *   packageId: 42,
   *   prefill: { firstName: "Jane", email: "jane@acme.com" },
   *   reminders: { enabled: true, intervalDays: 2 },
   * });
   * ```
   */
  async create(params: CreateSessionParams): Promise<CreateSessionResult> {
    const body: Record<string, unknown> = { packageId: params.packageId };
    if (params.prefill !== undefined)        body.prefill        = params.prefill;
    if (params.linkExpiryDays !== undefined) body.linkExpiryDays = params.linkExpiryDays;
    if (params.locale !== undefined)         body.locale         = params.locale;
    if (params.reminders !== undefined)      body.reminders      = params.reminders;
    if (params.signers !== undefined)        body.signers        = params.signers;

    return this.client.post<CreateSessionResult>("/sessions", body);
  }

  /**
   * Create up to 100 sessions in a single request.
   *
   * Each item in `sessions` accepts the same parameters as `create()`.
   * The batch processes all items — per-item failures do not abort the batch.
   * Check each result's `ok` field for success or failure.
   *
   * Returns HTTP 207 Multi-Status regardless of individual outcomes.
   *
   * @example
   * ```ts
   * const { results } = await client.sessions.bulkCreate({
   *   sessions: [
   *     { packageId: 42, prefill: { email: "alice@acme.com" } },
   *     { packageId: 42, prefill: { email: "bob@acme.com" } },
   *   ],
   * });
   * for (const r of results) {
   *   if (r.ok) console.log("Created:", r.sessionToken);
   *   else console.error("Failed item", r.index, ":", r.error);
   * }
   * ```
   */
  async bulkCreate(params: { sessions: BulkCreateSessionItem[] }): Promise<BulkCreateSessionResult> {
    return this.client.post<BulkCreateSessionResult>("/sessions/bulk", params);
  }

  /**
   * Fetch the current state of a session by its token.
   * Sensitive fields are redacted in the response.
   */
  async get(token: string): Promise<Session> {
    const res = await this.client.get<GetSessionResponse>(`/sessions/${token}`);
    return res.session;
  }

  /**
   * Retrieve the full chronological audit trail for a session.
   *
   * Each entry records a discrete event — session created, link sent, first
   * viewed, first answered, submitted, PDF generated, downloaded, voided, etc.
   * — including actor type, IP address, and structured metadata.
   *
   * Entries are returned oldest-first and are immutable once written.
   *
   * @example
   * ```ts
   * const { entries } = await client.sessions.auditLog(token);
   * for (const e of entries) {
   *   console.log(e.event, e.createdAt, e.actorType, e.actorIp);
   * }
   * ```
   */
  async auditLog(token: string, params?: { limit?: number }): Promise<AuditLogResult> {
    return this.client.get<AuditLogResult>(
      `/sessions/${token}/audit-log`,
      params?.limit !== undefined ? { limit: params.limit } : undefined,
    );
  }

  /**
   * Returns the ordered list of signers for a multi-party signing session.
   *
   * For sessions created with a `signers` array, this shows each signer's
   * current status and their unique interview token.
   */
  async signers(token: string): Promise<SessionSignersResult> {
    return this.client.get<SessionSignersResult>(`/sessions/${token}/signers`);
  }

  /**
   * Save interview answers for a session programmatically.
   * Useful when you want to fill the form on behalf of the client.
   */
  async updateAnswers(
    token: string,
    answers: Record<string, unknown>,
  ): Promise<Session> {
    const res = await this.client.patch<GetSessionResponse>(
      `/product/docuplete/sessions/${token}`,
      { answers },
    );
    return res.session;
  }

  /**
   * Trigger final PDF packet generation for a completed session.
   * Fires any enabled integrations (Google Drive, HubSpot, webhooks).
   *
   * When BullMQ is available the job is enqueued and `status === "pending"` is returned.
   * Poll `getGenerateStatus(token, jobId)` until `status === "ready"`.
   */
  async generate(token: string): Promise<GenerateSessionResult> {
    return this.client.post<GenerateSessionResult>(
      `/product/docuplete/sessions/${token}/generate`,
      {},
    );
  }

  /**
   * Poll the status of a background PDF generation job.
   *
   * @example
   * ```ts
   * const result = await client.sessions.generate(token);
   * if (result.status === "generated") {
   *   console.log("Ready:", result.downloadUrl);
   * } else {
   *   let ready = false;
   *   while (!ready) {
   *     await new Promise(r => setTimeout(r, 2000));
   *     const s = await client.sessions.getGenerateStatus(token, result.jobId);
   *     if (s.status === "ready") { console.log(s.downloadUrl); ready = true; }
   *     if (s.status === "failed") throw new Error(s.error ?? "Generation failed");
   *   }
   * }
   * ```
   */
  async getGenerateStatus(token: string, jobId?: string): Promise<GenerateStatusResult> {
    return this.client.get<GenerateStatusResult>(
      `/product/docuplete/sessions/${token}/generate-status`,
      jobId ? { jobId } : undefined,
    );
  }

  /**
   * Send (or re-send) the interview link email to a recipient.
   */
  async sendLink(token: string, params: SendLinkParams): Promise<{ ok: boolean; sentTo: string }> {
    return this.client.post<{ ok: boolean; sentTo: string }>(
      `/product/docuplete/sessions/${token}/send-link`,
      {
        recipientEmail: params.recipientEmail,
        recipientName:  params.recipientName,
        customMessage:  params.customMessage,
      },
    );
  }

  /**
   * Void a session, immediately invalidating its interview link.
   * Voided sessions cannot be submitted. This cannot be undone.
   */
  async void(
    token: string,
    params?: VoidSessionParams,
  ): Promise<{ ok: boolean; token: string; voidedAt: string }> {
    return this.client.post<{ ok: boolean; token: string; voidedAt: string }>(
      `/product/docuplete/sessions/${token}/void`,
      { reason: params?.reason, notifySigner: params?.notifySigner ?? false },
    );
  }
}

File: packages/python-sdk/docuplete/resources/sessions.py
"""Sessions resource — create, list, manage, and audit Docuplete sessions."""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from ..client import DocupleteClient
from ..types import (
    AuditLogResult,
    BulkCreateSessionItem,
    BulkCreateSessionResult,
    CreateSessionResult,
    GenerateSessionResult,
    GenerateStatusResult,
    ListSessionsParams,
    ListSessionsResult,
    RemindersConfig,
    SendLinkParams,
    Session,
    SessionSignerStatus,
    SessionSignersResult,
    SessionSigner,
    SessionStatus,
    SupportedLocale,
    VoidSessionParams,
    VoidSessionResult,
)


class SessionsResource:
    """
    Access via ``client.sessions``.

    Example
    -------
    >>> from docuplete import Docuplete
    >>> client = Docuplete(api_key="dp_live_...")
    >>> result = client.sessions.create(
    ...     package_id=42,
    ...     prefill={"first_name": "Jane", "email": "jane@example.com"},
    ...     link_expiry_days=14,
    ... )
    >>> print(result["interview_url"])
    """

    def __init__(self, client: DocupleteClient) -> None:
        self._client = client

    def list(
        self,
        *,
        package_id: Optional[int] = None,
        status: Optional[SessionStatus] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        updated_after: Optional[str] = None,
        search: Optional[str] = None,
    ) -> ListSessionsResult:
        """
        List sessions for your account with optional filters.

        Returns sessions in descending ``created_at`` order.

        Parameters
        ----------
        package_id : int, optional
            Filter to sessions belonging to a specific package.
        status : str, optional
            Filter by session status: ``"pending"``, ``"submitted"``,
            ``"generated"``, ``"voided"``, etc.
        limit : int, optional
            Maximum results to return (default: 50, max: 200).
        offset : int, optional
            Number of results to skip for pagination.
        updated_after : str, optional
            ISO-8601 timestamp — return only sessions updated after this time.
        search : str, optional
            Full-text search across prefill values.

        Example
        -------
        >>> sessions = client.sessions.list(status="generated", limit=25)
        >>> for s in sessions["sessions"]:
        ...     print(s["token"], s["status"])
        """
        params: Dict[str, Any] = {
            "packageId":    package_id,
            "status":       status,
            "limit":        limit,
            "offset":       offset,
            "updatedAfter": updated_after,
            "search":       search,
        }
        data = self._client.get("/sessions", params=params)
        return _snake(data)  # type: ignore[return-value]

    def create(
        self,
        *,
        package_id: int,
        prefill: Optional[Dict[str, str]] = None,
        link_expiry_days: Optional[int] = None,
        locale: Optional[SupportedLocale] = None,
        reminders: Optional[RemindersConfig] = None,
        signers: Optional[List[SessionSigner]] = None,
    ) -> CreateSessionResult:
        """
        Create a new interview session via the headless API.

        Returns a ``session_token`` and a ready-to-use ``interview_url``
        to send or redirect your client to.

        Prefill values should use field **source keys** as keys
        (e.g. ``{"first_name": "Jane", "email": "jane@example.com"}``).

        Parameters
        ----------
        package_id : int
            ID of the active package to use for this session.
        prefill : dict, optional
            Map of field source key → string value.
        link_expiry_days : int or None, optional
            Days until the link expires (1–3650). Pass ``None`` for a link
            that never expires. Omit to use your organisation default.
        locale : str, optional
            Interview language: ``"en"``, ``"es"``, ``"fr"``, etc.
        reminders : dict, optional
            ``{"enabled": True, "interval_days": 3}``
        signers : list, optional
            List of ``{"email": ..., "name": ..., "order": ...}`` dicts for
            multi-party sequential signing. Maximum 10 signers.

        Returns
        -------
        dict
            ``{"session_token": "df_...", "interview_url": "https://...", "expires_at": ...}``

        Example
        -------
        >>> result = client.sessions.create(
        ...     package_id=42,
        ...     prefill={"first_name": "Jane", "last_name": "Smith", "email": "jane@acme.com"},
        ...     link_expiry_days=14,
        ...     reminders={"enabled": True, "interval_days": 3},
        ... )
        >>> interview_url = result["interview_url"]
        >>> # Send interview_url to your client via email, SMS, or redirect
        """
        body: Dict[str, Any] = {"packageId": package_id}
        if prefill is not None:
            body["prefill"] = _camel_prefill(prefill)
        if link_expiry_days is not None:
            body["linkExpiryDays"] = link_expiry_days
        if locale is not None:
            body["locale"] = locale
        if reminders is not None:
            body["reminders"] = {
                "enabled":      reminders["enabled"],
                "intervalDays": reminders["interval_days"],
            }
        if signers is not None:
            body["signers"] = [
                {
                    "email": s["email"],
                    **({"name": s["name"]} if "name" in s else {}),
                    **({"order": s["order"]} if "order" in s else {}),
                }
                for s in signers
            ]

        data = self._client.post("/sessions", body=body)
        return _snake(data)  # type: ignore[return-value]

    def bulk_create(
        self,
        sessions: List[BulkCreateSessionItem],
    ) -> BulkCreateSessionResult:
        """
        Create up to 100 sessions in a single request.

        Each item accepts the same parameters as :meth:`create`. Per-item
        failures do not abort the batch — check each result's ``ok`` field.

        Returns HTTP 207 Multi-Status regardless of individual outcomes.

        Parameters
        ----------
        sessions : list
            Up to 100 session creation items.

        Example
        -------
        >>> contacts = [{"email": "a@co.com"}, {"email": "b@co.com"}]
        >>> result = client.sessions.bulk_create([
        ...     {"package_id": 42, "prefill": {"email": c["email"]}}
        ...     for c in contacts
        ... ])
        >>> for r in result["results"]:
        ...     if r["ok"]:
        ...         print("Created:", r["session_token"])
        ...     else:
        ...         print("Failed item", r["index"], ":", r["error"])
        """
        items = []
        for item in sessions:
            camel: Dict[str, Any] = {"packageId": item["package_id"]}
            if "prefill" in item:
                camel["prefill"] = _camel_prefill(item["prefill"])
            if "link_expiry_days" in item:
                camel["linkExpiryDays"] = item["link_expiry_days"]
            if "locale" in item:
                camel["locale"] = item["locale"]
            if "reminders" in item:
                r = item["reminders"]
                camel["reminders"] = {"enabled": r["enabled"], "intervalDays": r["interval_days"]}
            if "signers" in item:
                camel["signers"] = [
                    {"email": s["email"], **({} if "name" not in s else {"name": s["name"]})}
                    for s in item["signers"]
                ]
            items.append(camel)

        data = self._client.post("/sessions/bulk", body={"sessions": items})
        return _snake(data)  # type: ignore[return-value]

    def get(self, token: str) -> Session:
        """
        Fetch the current state of a session by its token.

        Parameters
        ----------
        token : str
            The session token (``df_...``).

        Example
        -------
        >>> session = client.sessions.get("df_a1b2c3...")
        >>> print(session["status"])   # "generated"
        >>> print(session["answers"])  # {"first_name": "Jane", ...}
        """
        data = self._client.get(f"/sessions/{token}")
        return _snake(data.get("session", data))  # type: ignore[return-value]

    def audit_log(
        self,
        token: str,
        *,
        limit: Optional[int] = None,
    ) -> AuditLogResult:
        """
        Retrieve the full chronological audit trail for a session.

        Each entry records a discrete event — session created, link sent, first
        viewed, submitted, PDF generated, downloaded, voided, etc. — with the
        actor type, IP address, and a UTC timestamp.

        Entries are returned oldest-first and are immutable once written.

        Example
        -------
        >>> log = client.sessions.audit_log("df_a1b2c3...")
        >>> for entry in log["entries"]:
        ...     print(entry["event"], entry["created_at"], entry["actor_ip"])
        """
        params: Dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        data = self._client.get(f"/sessions/{token}/audit-log", params=params or None)
        return _snake(data)  # type: ignore[return-value]

    def signers(self, token: str) -> SessionSignersResult:
        """
        Return the ordered list of signers for a multi-party signing session.

        For sessions created with a ``signers`` list, this shows each signer's
        current status, their unique interview token, and timestamps.

        Example
        -------
        >>> result = client.sessions.signers("df_a1b2c3...")
        >>> for signer in result["signers"]:
        ...     print(signer["email"], signer["status"])
        """
        data = self._client.get(f"/sessions/{token}/signers")
        return _snake(data)  # type: ignore[return-value]

    def generate(self, token: str) -> GenerateSessionResult:
        """
        Trigger final PDF generation for a completed session.

        When the job is queued, returns ``{"status": "pending", "job_id": "..."}``.
        Poll :meth:`get_generate_status` until ``status == "ready"``.

        Example
        -------
        >>> import time
        >>> result = client.sessions.generate("df_a1b2c3...")
        >>> if result["status"] == "generated":
        ...     print("Ready:", result["download_url"])
        ... else:
        ...     while True:
        ...         time.sleep(2)
        ...         s = client.sessions.get_generate_status(token)
        ...         if s["status"] == "ready":
        ...             print(s["download_url"])
        ...             break
        """
        data = self._client.post(
            f"/product/docuplete/sessions/{token}/generate", body={}
        )
        return _snake(data)  # type: ignore[return-value]

    def get_generate_status(
        self,
        token: str,
        job_id: Optional[str] = None,
    ) -> GenerateStatusResult:
        """
        Poll the status of a background PDF generation job.

        Parameters
        ----------
        token : str
            Session token.
        job_id : str, optional
            Job ID returned by :meth:`generate`. Pass for faster lookup.
        """
        params: Dict[str, Any] = {}
        if job_id is not None:
            params["jobId"] = job_id
        data = self._client.get(
            f"/product/docuplete/sessions/{token}/generate-status",
            params=params or None,
        )
        return _snake(data)  # type: ignore[return-value]

    def send_link(
        self,
        token: str,
        *,
        recipient_email: str,
        recipient_name: Optional[str] = None,
        custom_message: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Send (or re-send) the interview link email to a recipient.

        Example
        -------
        >>> client.sessions.send_link(
        ...     "df_a1b2c3...",
        ...     recipient_email="jane@example.com",
        ...     recipient_name="Jane Smith",
        ... )
        """
        body: Dict[str, Any] = {"recipientEmail": recipient_email}
        if recipient_name is not None:
            body["recipientName"] = recipient_name
        if custom_message is not None:
            body["customMessage"] = custom_message
        return self._client.post(  # type: ignore[return-value]
            f"/product/docuplete/sessions/{token}/send-link", body=body
        )

    def void(
        self,
        token: str,
        *,
        reason: Optional[str] = None,
        notify_signer: bool = False,
    ) -> VoidSessionResult:
        """
        Void a session, immediately invalidating its interview link.

        Voided sessions cannot be submitted. This action cannot be undone.

        Parameters
        ----------
        token : str
            Session token to void.
        reason : str, optional
            Reason for voiding — stored in the audit log.
        notify_signer : bool
            When ``True``, sends the signer an email notifying them the session
            was voided. Default: ``False``.

        Example
        -------
        >>> client.sessions.void("df_a1b2c3...", reason="Sent in error")
        """
        body: Dict[str, Any] = {"notifySigner": notify_signer}
        if reason is not None:
            body["reason"] = reason
        data = self._client.post(
            f"/product/docuplete/sessions/{token}/void", body=body
        )
        return _snake(data)  # type: ignore[return-value]


# ── Helpers ───────────────────────────────────────────────────────────────────

def _to_snake(s: str) -> str:
    """Convert a camelCase string to snake_case."""
    import re
    s = re.sub(r"([A-Z]+)([A-Z][a-z])", r"\1_\2", s)
    s = re.sub(r"([a-z\d])([A-Z])", r"\1_\2", s)
    return s.lower()


def _snake(obj: Any) -> Any:
    """Recursively convert all dict keys from camelCase to snake_case."""
    if isinstance(obj, dict):
        return {_to_snake(k): _snake(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_snake(item) for item in obj]
    return obj


def _camel_prefill(prefill: Dict[str, str]) -> Dict[str, str]:
    """
    Leave prefill keys as-is — they are field source keys defined by the user
    in the Docuplete dashboard and can be any case. Do not transform them.
    """
    return prefill
